\documentclass[12pt]{article} %lineno

\usepackage[top=0.8in,left=1in, right=1in, bottom=1in]{geometry}
\usepackage{graphicx}
\usepackage{hyperref}
\usepackage{multirow}
\usepackage{caption}
\usepackage{listing}

\usepackage{pgf}

\usepackage{alltt}

\hypersetup{colorlinks=true, linkcolor=blue, filecolor=blue, urlcolor=blue,
citecolor=blue }

\renewcommand{\figurename}{\textbf{Figure}}
\renewcommand{\thefigure}{\textbf{S\arabic{figure}}}

\renewcommand{\tablename}{\textbf{Table}}
\renewcommand{\thetable}{\textbf{S\arabic{table}}}

\renewcommand{\baselinestretch}{1.3}


% colors
\definecolor{constant}{HTML}{D55E00} % vermillion 
\definecolor{bluishgreen}{HTML}{009E73} % bluishgreen
\definecolor{blue}{HTML}{0072B2} % blue
\definecolor{generator}{HTML}{CC79A7} % reddishpurple
\definecolor{orange}{HTML}{E69F00} % orange
\definecolor{skyblue}{HTML}{56B4E9} % skyblue

% link colors
\hypersetup{
    colorlinks=true, 
    linkcolor=blue, 
    filecolor=blue, 
    urlcolor=blue,
    citecolor=blue
}


\begin{document}


\title{Supplementary Material for ``LinguaPhylo: a probabilistic model specification language for reproducible phylogenetic analyses''}

\date{}

\author{
Alexei J. Drummond\textsuperscript{1,2,3}*,
Kylie Chen\textsuperscript{1,2,3},
F\'{a}bio K. Mendes\textsuperscript{1,4},
Dong Xie\textsuperscript{1,2,3}
}

\maketitle 

{\small
\noindent \textbf{1} Centre for Computational Evolution, University of Auckland, Auckland, New Zealand
\\
\textbf{2} School of Biological Sciences, University of Auckland, Auckland, New Zealand
\\
\textbf{3} School of Computer Science, University of Auckland, Auckland, New Zealand
\\
\textbf{4} Department of Biology, Washington University in St. Louis, St. Louis, United States\\
}

\medskip

\noindent *Corresponding author: a.drummond@auckland.ac.nz

\clearpage


%\textbf{This PDF file includes:}

%\medskip 

%Figures XX to XX

%\medskip

%Tables XX to XX

% \newpage 

% \begingroup
% \pagestyle{plain}
% \listoffigures
\listoftables

\clearpage
% \endgroup

% \newpage

% Figures go here



% Tables

\begin{table}
\small
\centering
\begin{tabular}{ l | l | l }
    \hline\hline
    Function & Description & Examples \\ 
    \hline\hline
    binaryRateMatrix & Binary trait rate matrix & errorModel1.lphy, errorModel2.lphy\\  
    f81 & F81 model\cite{felsenstein1981} & f81Coalescent.lphy\\  
    generalTimeReversible & General time reversible rate matrix & h5n1.lphy \\  
    gtr & GTR model\cite{tarvare1986some} & gtrCoalescent.lphy\\  
    hky & HKY model\cite{hasegawa1985dating} & hkyCoalescent.lphy\\  
    jukesCantor & Jukes-Cantor model\cite{jc69} & jcCoalescent.lphy\\  
    k80 & K80 model\cite{kimura1980simple} & \\  
    lewisMK & LewisMK model\cite{lewis2001likelihood} & lewisMKCoalescent.lphy\\  
    migrationMatrix & Population process rate matrix & simpleStructuredCoalescent.lphy\\  
    wag & WAG model\cite{whelan2001general} & wagCoalescent.lphy\\  
    \hline
\end{tabular}

\label{tab:ratematrix}
\caption{Functions for substitution models and rate matrices in LPhy.}
\end{table}

\begin{table}
\small
\begin{tabular}{ l | l | l }
    \hline\hline
    Generative distribution & Description & Examples \\ 
    \hline\hline
    MultispeciesCoalescent & Multispecies coalescent & simpleMultispeciesCoalescent.lphy, \\ & & simpleMultispeciesCoalescentTaxa.lphy, \\ & & twoGeneMultispeciesCoalescent.lphy\\  
    Coalescent & Kingman's coalescent \cite{Rodrigo1999SerialCoalescent} & RSV2.lphy \\  
    SkylineCoalescent & Skyline coalescent \cite{drummond2005bayesiansequences} & hcv\_col.lphy\\  
    StructuredCoalescent & Structured coalescent\cite{muller2017structured} & simpleStructuredCoalescent.lphy\\  
    \hline
\end{tabular}

\label{tab:coalescent}
\caption{Coalescent tree generative distributions in LPhy.}
\end{table}

\begin{table}
\small
\begin{tabular}{ p{4.6cm} | p{6.5cm} | p{5cm} }
    \hline\hline
    Generative distribution & Description & Examples \\ 
    \hline\hline
    BirthDeathSampling & Birth-death-sampling tree\cite{stadler2011mammalian,stadler2012estimatingdata} & birthDeathRhoSampling.lphy\\  
    BirthDeathSerialSampling & Birth-death serial sampling tree\cite{stadler2013dating} & simpleBirthDeathSerial.lphy\\  
    BirthDeath & Calibrated birth-death\cite{heled2015calibrated} & simpleCalibratedBirthDeath.lphy, \\ & & simpleExtantBirthDeath.lphy\\  
    FossilBirthDeathTree & Fossilized birth-death process\cite{heath2014fossilized} & simFossilsCompact.lphy\\  
    FullBirthDeath & Birth-death tree\cite{kendall1948generalized} & simpleFullBirthDeath.lphy\\  
    RhoSampleTree & Birth-death tree sampled from a larger tree & \\  
    SimBDReverse & Birth-death tree with extant and extinct species & simFossils.lphy\\  
    SimFBDAge & Birth-death tree with extant and extinct species sampled through time & simFBDAge.lphy \\   
    SimFossilsPoisson & Tree with fossils added to given tree at rate psi & simFossils.lphy\\  
    Yule & Yule tree\cite{yule1925ii} & simpleYule.lphy, \\ 
    & & yuleRelaxed.lphy\\  
    \hline
\end{tabular}
\label{tab:coalescent}
\caption{Birth-death tree generative distributions in LPhy.}
\end{table}

\begin{table}
\small
\begin{center}
\begin{tabular}{ l | l | l }
    \hline\hline
    Generative distribution & Description & Examples \\ 
    \hline\hline
    PhyloBrownian & Brownian motion process\cite{felsenstein1973maximum} & simplePhyloOU.lphy\\  
    PhyloCTMC & Continuous time Markov process\cite{felsenstein1981} & simpleBModelTest.lphy\\  
    PhyloMultivariateBrownian & Multivariate Brownian motion & simplePhyloMultivariateBrownian.lphy\\  
    PhyloOU & Ornstein-Ulhenbeck process\cite{felsenstein1973maximum} & simplePhyloBrownian.lphy\\  
    \hline
\end{tabular}
\end{center}
\label{tab:coalescent}
\caption{Phylogenetic likelihood distributions in LPhy.}
\end{table}

\begin{table}
\small
\begin{tabular}{ l | p{7cm} | l }
    \hline\hline
    Generative distribution & Description & Examples \\ 
    \hline\hline
    Bernoulli & Coin toss distribution & simpleRandomLocalClock.lphy, \\ & & simpleBModelTest.lphy\\  
    Beta & Beta distribution & birthDeathRhoSampling.lphy, \\ & & simpleBModelTest.lphy\\  
    Cauchy & Cauchy distribution & \\  
    Dirichlet & Dirichlet distribution & birthDeathRhoSampling.lphy, \\ & & dirichlet.lphy\\  
    DiscreteUniform & Discrete-uniform distribution & simpleBModelTest.lphy, \\ & & simpleBModelTest2.lphy\\  
    DiscretizeGamma &  Discretize-gamma distribution & gtrGammaCoalescent.lphy, \\ & & simpleBModelTest.lphy\\  
    Exp & Exponential distribution & birthDeathRhoSampling.lphy, \\ & & yuleRelaxed.lphy\\  
    ExpMarkovChain & Smoothing distribution \cite{drummond2005bayesiansequences} & skylineCoalescent.lphy\\  
    Gamma & Gamma distribution & covidDPG.lphy\\  
    Geometric & Geometric distribution & \\  
    InverseGamma & Inverse-gamma  distribution & totalEvidence.lphy\\  
    LogNormal & Log-normal distribution & hkyCoalescent.lphy, \\ & & errorModel1.lphy\\  
    Normal & Normal distribution & simplePhyloBrownian.lphy, \\ & & simplePhyloOU.lphy\\  
    NormalGamma & Normal-gamma distribution & simplePhyloBrownian.lphy, \\ & & simplePhyloOU.lphy\\  
    Poisson & Poisson distribution & expression4.lphy, \\ & & simpleRandomLocalClock2.lphy\\  
    RandomBooleanArray & Samples a random boolean array & simpleRandomLocalClock2.lphy\\  
    RandomComposition & Samples a random k-tuple of positive integers that sum to n & skylineCoalescent.lphy\\  
    Uniform & Uniform distribution & simFossilsCompact.lphy\\  
    Weibull & Weibull distribution & \\  
    WeightedDirichlet & Weighted dirichlet distribution & totalEvidence.lphy, \\ & & weightedDirichlet.lphy\\  
    \hline
\end{tabular}

\label{tab:prior}
\caption{Parametric distributions in LPhy.}
\end{table}


\begin{table}
\begin{center}
\small
\begin{tabular}{ l | l | l }
    \hline\hline
    Function & Description & Examples \\ 
    \hline\hline
    aminoAcids & Amino acid data type & wagCoalescent.lphy\\  
    binaryDataType & Binary data type & \\  
    nucleotides & Nucleotide data type & primates2.lphy\\  
    standard & Standard data type & totalEvidence.lphy\\  
    \hline
\end{tabular}
\label{tab:coalescent}
\caption{Alignment data types in LPhy.}
\end{center}
\end{table}


\begin{table}
\small
\begin{tabular}{ l | l | l }
    \hline\hline
    Function & Description & Examples \\ 
    \hline\hline
    nucleotideModel & bModelTest\cite{bouckaert2017bmodeltestcomparison} rate matrix & simpleBModelTest.lphy, \\ & & simpleBModelTest2.lphy\\  
    bModelSet & bModelTest model set & simpleBModelTest.lphy\\  
    bSiteRates & Site rates for the given bModelTest parameters & simpleBModelTest2.lphy\\  
    bSiteModel & bModelTest site model & simpleBModelTest.lphy\\  
        \hline
\end{tabular}
\label{tab:coalescent}
\caption{Bayesian phylogenetic site model averaging in LPhy.}
\end{table}

\clearpage

\bibliographystyle{plain}

\bibliography{linguaPhylo}


\end{document}